op-node/rollup/derive/channel_stage.go

```diff
@@ -0,0 +1,120 @@
+package derive
+
+import (
+	"context"
+	"io"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type ChannelStage struct {
+	log     log.Logger
+	spec    *rollup.ChainSpec
+	metrics Metrics
+
+	channel *Channel
+
+	prev NextFrameProvider
+}
+
+var _ ResettableStage = (*ChannelStage)(nil)
+
+// NewChannelStage creates a Holocene ChannelStage.
+// It must only be used for derivation from Holocene activation.
+func NewChannelStage(log log.Logger, cfg *rollup.Config, prev NextFrameProvider, m Metrics) *ChannelStage {
+	return &ChannelStage{
+		log:     log,
+		spec:    rollup.NewChainSpec(cfg),
+		metrics: m,
+		prev:    prev,
+	}
+}
+
+func (cs *ChannelStage) Log() log.Logger {
+	return cs.log.New("stage", "channel", "origin", cs.Origin())
+}
+
+func (cs *ChannelStage) Origin() eth.L1BlockRef {
+	return cs.prev.Origin()
+}
+
+func (cs *ChannelStage) Reset(context.Context, eth.L1BlockRef, eth.SystemConfig) error {
+	cs.resetChannel()
+	return io.EOF
+}
+
+func (cs *ChannelStage) resetChannel() {
+	cs.channel = nil
+}
+
+// Returns whether the current staging channel is timed out. Panics if there's no current channel.
+func (cs *ChannelStage) channelTimedOut() bool {
+	return cs.channel.OpenBlockNumber()+cs.spec.ChannelTimeout(cs.Origin().Time) < cs.Origin().Number
+}
+
+func (cs *ChannelStage) NextData(ctx context.Context) ([]byte, error) {
+	if cs.channel != nil && cs.channelTimedOut() {
+		cs.metrics.RecordChannelTimedOut()
+		cs.resetChannel()
+	}
+
+	lgr := cs.Log()
+	origin := cs.Origin()
+
+	// Note that if the current channel was already completed, we would have forwarded its data
+	// already. So we start by reading in frames.
+	if cs.channel != nil && cs.channel.IsReady() {
+		panic("unexpected ready channel")
```

Comment:
nit: Should we panic here or return a FatalError?

Reply:
It would be a logic error local to this function if the channel was ready here, so I used a proper panic. Changed now to a `NewCriticalError`. But since this would really hint at a programming error, I would still prefer a panic. It's like a runtime assertion of a program invariant. Let me know what you think.

Reply:
I'm not really sure what the right answer is here to be honest. I just raised it because I don't recall seeing `panic` through the derivation pipeline code much and `CriticalError` is used for things that are unrecoverable and shouldn't ever happen. Both will wind up exiting the program with an error exit code. The main difference seems to be that panic will just print to stderr and CriticalError will report the error through the logging framework.

I'm happy to go with your call here.  Not sure if @protolambda has an opinion.

Reply:
Since the effect is the same, I will stick with the `CriticalError`.